### PR TITLE
Add Hindley-Milner type inference module

### PR DESCRIPTION
**Pull request by samarsha**
This PR splits out the HM type inference module from #884, as well as additional changes that the module requires. The new module is not used by the compiler, so there's no change to type inference behavior yet (there will be another PR for that).

Additional changes:

* Added a range field to `ResolvedType` so that the HM module can generate diagnostics with accurate locations.
* Split ScopeTools.fs into two files so that the HM module can reference SymbolTracker, and ScopeContext can reference the HM module.
* Removed `WithinIfCondition` field from `ScopeContext` that has been deprecated for 6 months.

### CODE REVIEW

File: src/QsCompiler/SyntaxProcessor/ScopeContext.fs
Comment (swernli) [inline src/QsCompiler/SyntaxProcessor/ScopeContext.fs:34]: I know this was there in the previous version and you are just preserving it, but why do we need the processor architecture for syntax processing?

**Review by swernli — APPROVED** [review]
Changes look good! The InferenceContext.fs file has a lot of logic that is hard to internalize when coming at the code fresh, so I'm wondering if it might be worth having a readme in the TypeInference folder with some brief descriptions of the high level concepts/strategy and a link to external info on Hindley-Milner type inference.

File: src/QsCompiler/SyntaxProcessor/ScopeContext.fs
Comment (samarsha) [reply src/QsCompiler/SyntaxProcessor/ScopeContext.fs:34]: It's used only for diagnostic messages, so they can include the name of the quantum processor in capability errors.

Comment (swernli) [reply src/QsCompiler/SyntaxProcessor/ScopeContext.fs:34]: Ah, ok. That does make sense. I have been so caught up in working on clang, LLVM, and CMakeLists.txt setup that I misinterpreted this as the classical processor architecture that the code was compiled on, which for Q# didn't seem relevant at all. Thanks for explaining!

**Comment by samarsha** [comment]
@swernli I added a readme. Can you take a look at it?

File: src/QsCompiler/DataStructures/Diagnostics.fs
Comment (cesarzc) [inline src/QsCompiler/DataStructures/Diagnostics.fs:10]: I wonder if errors starting at 1001 was a conscious decision to show all Q# compilation errors look similar (e.g. _error QS2003_). Would it make sense to start type errors at 4001?

Comment (samarsha) [reply src/QsCompiler/DataStructures/Diagnostics.fs:10]: Yeah, the error codes I picked are unusual compared to the existing ones. I think there are two separate questions:

1. Should error codes start at 0001 or 1001?
2. Are there groups (first digit, first two digits, etc.) of error codes for different kinds of errors?

For question 1, I would rather format error codes with leading zeroes than require that all error codes start at 1001, to avoid error code inflation. For comparison, [FS0001](https://docs.microsoft.com/en-us/dotnet/fsharp/language-reference/compiler-messages/fs0001) is the F# error code for type mismatch.

For question 2, I'm not sure that there is a consistent grouping of error codes by leading digits. For example, ExpectingIntExpr and ArgumentTypeMismatch are both type errors, but ExpectingIntExpr is QS5012 and ArgumentTypeMismatch is QS6210. ExpectingSemicolon and ExpectingIdentifier are both syntax errors, but ExpectingSemicolon is QS2003 and ExpectingIdentifier is QS6103.

**Comment by swernli** [comment]
@samarsha The readme is fantastic! Concise and still very informative for interpreting the accompanying code.

File: src/QsCompiler/DataStructures/Diagnostics.fs
Comment (cesarzc) [reply src/QsCompiler/DataStructures/Diagnostics.fs:10]: I agree with you on both answers. Just before closing this conversation, with this change if there is a TypeMismatch error would the user see QS0001? When the error is formatted to be shown to the user is it adding the leading zeroes?

Comment (samarsha) [reply src/QsCompiler/DataStructures/Diagnostics.fs:10]: If there is a pattern to error code number groups, then I think it would help to have comments that describe each group, since I don't think that they are very clear from context. I think @bettinaheim might know what the groups are?

Comment (samarsha) [reply src/QsCompiler/DataStructures/Diagnostics.fs:10]: > Just before closing this conversation, with this change if there is a TypeMismatch error would the user see QS0001? When the error is formatted to be shown to the user is it adding the leading zeroes?

Currently, it's formatted as QS1, but I can change it now to add leading zeroes.

File: src/QsCompiler/SyntaxProcessor/TypeInference/Constraint.fs
Comment (cesarzc) [inline src/QsCompiler/SyntaxProcessor/TypeInference/Constraint.fs:9]: I am a curious about why the `Constraint` type is defined in two separate source files?

Comment (samarsha) [reply src/QsCompiler/SyntaxProcessor/TypeInference/Constraint.fs:9]: Constraint.fsi is the signature file for Constraint.fs. In general, the signature file contains only the types of things in the implementation file (so Constraint.fsi has `val types: Constraint -> ResolvedType list`, while Constraint.fs has `let types = ...`).

For discriminated unions and records with public or internal cases/fields, the signature and implementation are actually the same, but it's still required to provide them separately, so unfortunately they are duplicated across both files.

File: src/QsCompiler/DataStructures/Diagnostics.fs
Comment (samarsha) [reply src/QsCompiler/DataStructures/Diagnostics.fs:10]: It's formatted with leading zeroes now (QS0001).

File: src/QsCompiler/SyntaxProcessor/TypeInference/InferenceContext.fsi
Comment (bettinaheim) [inline src/QsCompiler/SyntaxProcessor/TypeInference/InferenceContext.fsi:33]: Hm, what confuses me a bit is why you don't have a method for unification that computes the more general type of the two, and one that computes the less general type of the two ("covariant vs contravariant unification"). How does that work currently?

File: src/QsCompiler/Core/Transformations/TypeTransformation.fs
Comment (bettinaheim) [inline src/QsCompiler/Core/Transformations/TypeTransformation.fs:147]: Please make sure the order in which individual items in a node are processed matches other transformations; For the other nodes, ranges are processed before the kinds and the types.

**Review by bettinaheim — CHANGES_REQUESTED** [review]
I am blocking the PR for now since I believe in its current state it would break some of the IDE commands due to the change in assigning ranges to types even when there is no matching piece of source code.

File: src/QsCompiler/SyntaxProcessor/TypeInference/InferenceContext.fsi
Comment (samarsha) [reply src/QsCompiler/SyntaxProcessor/TypeInference/InferenceContext.fsi:33]: That is `UnifyByOrdering`.

https://github.com/microsoft/qsharp-compiler/blob/70f506f9ad6b14554572bcdccba4485a4daf821c/src/QsCompiler/SyntaxProcessor/TypeInference/InferenceContext.fs#L362-L366

File: src/QsCompiler/SyntaxProcessor/SymbolTracker.fs
Comment (samarsha) [reply src/QsCompiler/SyntaxProcessor/SymbolTracker.fs]: To clarify based on our other conversation, the PR didn't cause issues with IDE commands since it left alone the ranges in QsTypeParameter and UserDefinedType. However, having a range for types in multiple places was strange, so I updated the range in ResolvedType to distinguish between an inferred type and an annotated type, and marked the ranges for TPs and UDTs as obsolete.

Comment (samarsha) [reply src/QsCompiler/SyntaxProcessor/SymbolTracker.fs]: Currently, the IDE commands are still using the obsolete ranges instead of the new ResolvedType.Range, but I'd like to update that in another PR to avoid making this one too big.

File: src/QsCompiler/DataStructures/SyntaxTree.fs
Comment (bettinaheim) [inline src/QsCompiler/DataStructures/SyntaxTree.fs:407]: nit: TryGetRange to me is a bit more comprehensive than TryRange.

File: src/QsCompiler/Core/Transformations/TypeTransformation.fs
Comment (bettinaheim) [inline src/QsCompiler/Core/Transformations/TypeTransformation.fs:126]: Hm, to make the change in behavior clearer, I would deprecate OnRangeInformation, and instead make separate methods to overload for annotated and inferred ranges. I think that will make it clearer that and how transformations need to be updated. Do you see any downsides to that?

File: src/QsCompiler/SyntaxProcessor/TreeVerification.fs
Comment (bettinaheim) [reply src/QsCompiler/SyntaxProcessor/TreeVerification.fs:6]: Finally my NuGet packages have restored and I can build things again locally; It looks like this can be removed.

**Review by bettinaheim — COMMENTED** [review]
I think this setup make things a lot cleaner, thanks. I am still waiting for a response regarding SearchAndReplace; we need to update the transformation to not use the deprecated handles. The syntax tree transformation currently walks both the old and the new handle, which I believe would cause issues.

File: src/QsCompiler/BondSchemas/BondSchemaTranslator.cs
Comment (bettinaheim) [inline src/QsCompiler/BondSchemas/BondSchemaTranslator.cs:539]: Actually, why  do we need to keep this? Isn't this what gets serialized; i.e. can't we drop this and just make sure that the corresponding item in already compiled old dlls is ignored when deserializing it?

File: src/QsCompiler/DataStructures/SyntaxTree.fs
Comment (bettinaheim) [inline src/QsCompiler/DataStructures/SyntaxTree.fs:136]: I saw that in the NamespaceManager there are a lot of usages of this now deprecated handle. They are not showing up here because the file already contains a nowarn due to the access modifier changes. Please take a look at what updates are required there - I believe it should be possible to update it entire to use only the new handles.

File: src/QsCompiler/Tests.Compiler/SerializationTests.fs
Comment (bettinaheim) [reply src/QsCompiler/Tests.Compiler/SerializationTests.fs:7]: Looking more closely now that I can build locally, I think it would be better to just update the one test that is there and remove this.

File: src/QsCompiler/Tests.Compiler/TestUtils/Signatures.fs
Comment (bettinaheim) [reply src/QsCompiler/Tests.Compiler/TestUtils/Signatures.fs:6]: Same here; the tests can easily be updated to not use the old handles.

File: src/QsCompiler/Transformations/SearchAndReplace.cs
Comment (bettinaheim) [reply src/QsCompiler/Transformations/SearchAndReplace.cs:233]: After looking through all code, I realize that you maintain the full support for the old handles and they are still populated (that makes sense, thank you!). Regardless, we need to update everything within the compiler to work only with non-deprecated items, otherwise this will just haunt us once we remove the support for the deprecated handles.

File: src/QsCompiler/SyntaxProcessor/SymbolTracker.fs
Comment (bettinaheim) [reply src/QsCompiler/SyntaxProcessor/SymbolTracker.fs]: Ok, I am good with making it a separate PR, so I'll unblock the PR. However, please address the comment around removing other suppressions of deprecated warnings where possible (they are all small to update).

File: src/QsCompiler/Transformations/ContentLifting.cs
Comment (samarsha) [reply src/QsCompiler/Transformations/ContentLifting.cs:442]: It doesn't have access to `ResolvedType.Range` because the parameter its given is only a `QsTypeParameter`. The update is non-trivial, so I will do it in a different PR.

File: src/QsCompiler/SyntaxProcessor/TypeInference/InferenceContext.fsi
Comment (samarsha) [reply src/QsCompiler/SyntaxProcessor/TypeInference/InferenceContext.fsi:34]: I updated the doc comments for `Unify` and `Intersect`. For `Unify`, `actual` must be a subtype of `expected`. `Intersect` returns a type that is a supertype of the given types.

The reason that the more general functions are private is simply because there is no reason for any other part of the compiler to ask for an invariant or contravariant unification or intersection. That only happens when recursing into an array or callable type, which only occurs in the inference module itself.

File: src/QsCompiler/DataStructures/SyntaxTree.fs
Comment (samarsha) [reply src/QsCompiler/DataStructures/SyntaxTree.fs:412]: Do you mean for the compiled name, the F# name, or both?

File: src/QsCompiler/BondSchemas/BondSchemaTranslator.cs
Comment (samarsha) [reply src/QsCompiler/BondSchemas/BondSchemaTranslator.cs:539]: Well, we could drop the ranges during serialization, but that is technically a breaking change if anyone is relying on them in a deserialized syntax tree (e.g. a rewrite step that looks at references in the AST). Probably no one would be affected by it, but I wanted to continue fully supporting the old ranges until they are removed.

Comment (samarsha) [reply src/QsCompiler/BondSchemas/BondSchemaTranslator.cs:539]: Or do you mean that since the old behavior was to strip range information during deserialization, then this wouldn't be a breaking change?

File: src/QsCompiler/DataStructures/SyntaxTree.fs
Comment (samarsha) [reply src/QsCompiler/DataStructures/SyntaxTree.fs:136]: I'm not sure what you mean by a pre-existing nowarn for access modifiers? There is one in Namespace.fs but not NamespaceManager.fs. I added the nowarn to NamespaceManager.fs deliberately, because it relies on the range of UserDefinedType for attribute diagnostics. I think fixing that is a non-trivial change that I would like to do in a separate PR.

File: src/QsCompiler/Core/Transformations/TypeTransformation.fs
Comment (samarsha) [reply src/QsCompiler/Core/Transformations/TypeTransformation.fs]: I think adding separate abstract methods for each case of `TypeRange` is too much, but I did deprecate `OnRangeInformation` and added a new `OnTypeRange`.

File: src/QsCompiler/DataStructures/SyntaxTree.fs
Comment (bettinaheim) [reply src/QsCompiler/DataStructures/SyntaxTree.fs:412]: Both, but just a thought.

File: src/QsCompiler/Core/Transformations/TypeTransformation.fs
Comment (bettinaheim) [reply src/QsCompiler/Core/Transformations/TypeTransformation.fs]: Ok, that works for me.

File: src/QsCompiler/BondSchemas/BondSchemaTranslator.cs
Comment (bettinaheim) [reply src/QsCompiler/BondSchemas/BondSchemaTranslator.cs:539]: I think this might not actually break anything if we remove it, and people are supposed to be using the compiler methods to load the syntax tree rather than rely on the serialization direction. So as long as that keeps working for old dlls, we are good to remove this.

File: src/QsCompiler/DataStructures/SyntaxTree.fs
Comment (bettinaheim) [reply src/QsCompiler/DataStructures/SyntaxTree.fs:136]: Yes, my desire was to update the diagnostics generation. Making this a separate PR makes sense.

File: src/QsCompiler/DataStructures/SyntaxTree.fs
Comment (samarsha) [reply src/QsCompiler/DataStructures/SyntaxTree.fs:412]: I don't think that's consistent with F# naming conventions; compare with `List.tryHead`, `List.tryLast`, etc. I generally think "get" is overused in function names since most functions are getting something.

**Pull request merged**
